Signing templates

DonKoko opened the pull request.

DonKoko commented on app/atoms/file.ts:
```diff
@@ -1,8 +1,12 @@
+fileErrorAtom.onMount = (setAtom) => {
+  setAtom("");
+};
+export const MAX_SIZE = 4_000_000; // 4MB
```
I think we need a bit more size for the PDF's. Lets say 8MB.

rajdip-b replied:
Resolved by [d18e0fc](https://github.com/Shelf-nu/shelf.nu/pull/719/commits/d18e0fc911abaff21fd630a074fa1829327726be)

DonKoko commented on app/components/sign/agreement-popup.tsx:
```diff
@@ -0,0 +1,53 @@
+  return (
+    <AnimatePresence>
+      {isMobileView ? (
+        <div
```
Why not use the Dialog component we have already?

rajdip-b replied:
I copied this one from the asset page :/

rajdip-b replied:
@DonKoko can you link the usage of the existing component once?

DonKoko replied:
@rajdip-b here is the component: https://github.com/Shelf-nu/shelf.nu/blob/main/app/components/layout/dialog.tsx
If you search for `<Dialog` you should be able to see how we use it, but its pretty straight forward.

DonKoko commented on app/components/sign/agreement.tsx:
```diff
@@ -0,0 +1,88 @@
+            penColor="gray"
+            canvasProps={{
+              className: "sigCanvas",
+              // @TODO NEeds to be fixed @rajdip-b. I dont see anywhere in the docs a placeholder prop
```
@rajdip-b left a comment in the code

rajdip-b replied:
don't know what i was thinking :)

rajdip-b replied:
Resolved by [d18e0fc](https://github.com/Shelf-nu/shelf.nu/pull/719/commits/d18e0fc911abaff21fd630a074fa1829327726be)

DonKoko commented on app/components/templates/form.tsx:
```diff
@@ -0,0 +1,285 @@
+          disabled={disabled}
+          onValueChange={(val) =>
+            setSelectedType(
+              val.toString() === TemplateType.BOOKINGS
```
Maybe we should drop bookings for now so we can release custody first without being blocked by the bookings signature.

rajdip-b replied:
uh, so what should i change in this one?

DonKoko replied:
We have bookings now so this comment is invalid :)

DonKoko commented on app/modules/asset/service.server.ts:
```diff
@@ -244,6 +246,12 @@ export async function getAssets({
             },
             custody: {
               select: {
+                templateSigned: true,
```
I am not sure why we need that in `getAssets`.  Are we showing this data anywhere on an index?

rajdip-b replied:
I'm not entirely sure I get that question. If you mean the `templateSigned` field or the custody, I am using that to display the "SIgnature pending" logic

rajdip-b replied:
@DonKoko please review this logic once so that I can mark its status

DonKoko replied:
@rajdip-b `getAssets` is used in the index as well as some other places. Including the templateSigned in all of those will require an extra join and will make the query slower in places where we dont use the data. We need to either allow getAssets & getPaginatedAndFiltered assets to receive a dynamic include or we need to make a separate function for this one case.

DonKoko commented on app/modules/template/service.server.ts:
```diff
@@ -0,0 +1,240 @@
+   * the template via mailing them.
+   */
+  if (signatureRequired === true) {
+    const custodies = await db.custody.findMany({
```
This is a very interesting case. But I am not sure this is the best way to handle this. 
@jurrejansen @carlosvirreira we need to take some decisions here. Here is the case:
1. Template exists and doesnt require signature and some people already have custodies assigned through that template.
2. The template gets update to signature required

What do we do? I think we discussed something that we just dont allow templates to be updated after they have 1 custody using them. What do we think?

DonKoko replied:
@carlosvirreira @jurrejansen what do you guys think about this? We discussed that if a template is used even once, we should not allow for the file to be changed, but should we consider this also for the case of changing required? I would say yes.

DonKoko commented on app/modules/template/service.server.ts:
```diff
@@ -0,0 +1,240 @@
+  return updateTemplate;
+}
+
+export async function updateTemplatePDF({
```
@jurrejansen @rajdip-b @carlosvirreira this is the same case. We have to consider weather we want to allow this. I think once a template has been attached to even 1 custody, we should not allow changing of the PDF.

rajdip-b replied:
I agree. This will reduce a lot of problems. Plus, me and @carlosvirreira also decided on (i think) to not let the users update the template PDF or any other settings once it is associated to a custody (ASSIGNED)

DonKoko replied:
Agreed. Let's implement it like that. Once a template has been used even once, we don't allow it to be edited anymore. They have to make a new one.

DonKoko commented on app/modules/template/service.server.ts:
```diff
@@ -0,0 +1,240 @@
+    organizationId,
+  };
+
+  const [templates, totalTemplates] = await Promise.all([
```
Just curious, why not use transactions from prisma?

rajdip-b replied:
I was not familiar with prisma tx back then, will do it now

rajdip-b replied:
Resolved

DonKoko replied:
OKey I am sorry, but we have to reverse that. In one of the latest things we did, we basically did some performance testing and Promise.all is 30-70% faster than using prisma's transaction. The change should be pretty easy tho

DonKoko reviewed (commented):
- [ ] When the user selects a custody template that doesn't require a signature, in the notification email we should send the custody agreement file via attachment
- [ ] When we are waiting for signature, we should also show a little card on the asset page, not only the icon on top: https://www.figma.com/file/ykvc0MhPeNLsrJ4tJsKosQ/Shelf-%7C-UX-Scenarios?type=design&node-id=2109-61476&mode=design&t=1eL6hsp7kdUUCA9r-0
![Screenshot 2024-01-30 at 15 29 03](https://github.com/Shelf-nu/shelf.nu/assets/7840007/a969070b-7bf4-4cf1-bbf1-8629bf521e88)
- [ ] When a user signs a template to receive custody, we should be able to see that somewhere. I can see there is a design for it but I cant find it in the app https://www.figma.com/file/ykvc0MhPeNLsrJ4tJsKosQ/Shelf-%7C-UX-Scenarios?type=design&node-id=626-29627&mode=design&t=qbrdFTcqJzquhfbK-0
- [ ] After signing a custody assignment for an asset, i can click back and go back to the signing page. This should not be possible. Once signed and the user lands on the page, we should throw an error that this is already signed
- [ ] When I try to access a signing url with a different user it gives an error, which is good, but the error message is not great:
![Screenshot 2024-01-30 at 15 44 57](https://github.com/Shelf-nu/shelf.nu/assets/7840007/71496c07-a566-474b-8829-801033376cd5)
```
Title: Unauthorized
Text: You are not authorized to view this page and sign this template. Please make sure you are logged in with the correct account.
```
- [ ] If you click a sign url and you are not in the correct workspace, you get an error "Asset not found". We need to think of a way to manage this. Maybe what we can do is just get the asset 
- [ ] We have to think about the bookings signing. Maybe we just disabled the "bookings" type for templates for now until we figure it out.
- [ ] When i try to give custody to an available asset to a person that already has custody of another asset using the same template
![Screenshot 2024-01-30 at 16 30 14](https://github.com/Shelf-nu/shelf.nu/assets/7840007/a0f38d64-7513-4757-88d1-736fdc0b0e4c)

DonKoko commented on app/routes/_layout+/assets.$assetId.share-template.tsx:
I have no idea how to access this view from the UI.

DonKoko commented on app/routes/_layout+/assets._index.tsx:
```diff
@@ -343,11 +347,38 @@ const ListAssetContent = ({
+                    <CustomTooltip
+                      content={
+                        <div className="flex flex-col gap-y-2 p-3">
+                          <span className="text-sm text-gray-700">
```
I adjusted how the content is structured inside assetId, Can you also please adjust it here @rajdip-b

rajdip-b replied:
Can you elaborate on this one? not sure what i need to change :/

DonKoko replied:
Just making sure the custom tooltip content is the same. It was quite some time ago so I dont remember exactly what changed. Just make sure the tooltip content is consistent with assetId route

DonKoko commented on app/routes/_layout+/settings.template.$templateId_.edit.tsx:
I think here we have to implement some check, we have some cases when updating a template should not be allowed. I already asked for opinions in another comment.

rajdip-b replied:
yeap

DonKoko commented on app/routes/_layout+/settings.template.index.tsx:
```diff
@@ -0,0 +1,293 @@
+  const { organizationId } = await requireOrganisationId(authSession, request);
+  const { userId } = authSession;
+
+  const user = await db.user.findUnique({
```
I am not sure why we are doing this here. You can get userId from session and then query templates by themself. Should be faster.

rajdip-b replied:
Resolved

DonKoko commented on app/routes/_sign+/_sign.tsx:
I think its way more clear if we name this file _layout

rajdip-b commented:
@DonKoko cool, i'll go through them and start fixing them one by one

DonKoko commented on app/utils/storage.server.ts:
```diff
@@ -18,6 +19,8 @@ export function getPublicFileURL({
 }) {
+  bucketExists(bucketName);
+
   const { data } = getSupabaseAdmin()
```
Shouldnt this be async?

rajdip-b replied:
Yeah! missed it

rajdip-b replied:
Resolved

DonKoko reviewed (commented):
Submitted the first review by mistake so I had to do a second one for the rest of the files.

rajdip-b commented on app/routes/_sign+/sign.$templateId.tsx:
```diff
@@ -0,0 +1,273 @@
+import tailwindConfig from "../../../tailwind.config";
+
+export const loader = async ({ request }: LoaderFunctionArgs) => {
+  const authSession = await requireAuthSession(request);
```
Resolved

DonKoko commented:
@rajdip-b okey here is the update:

1. I have updated `signing-templates` with `main`. So you need to update your branch with `signing-templates`
2. I have fixed the conflicts and added a few `@TODO` comments that need to be addressed before we merge this
3. Run `npm run typecheck` and you will see a bunch of TS errors. That is because you have some files and routes that dont follow the new route error/response handling. Those routes need to be updated. Here are the docs: https://github.com/Shelf-nu/shelf.nu/blob/main/docs/handling-errors.md

carlosvirreira commented:
hi @rajdip-b was the update from Niko helpful?

DonKoko commented:
@rajdip-b Updated the PR with main. 
Considering what you mentioned on discord, that you already fixed the loaders and actions, we will most likely have conflicts again as I had to touch a lot of files that had issues because of the release of Vite.
Let me know when you push your changes so I can help with fixing the conflicts.

DonKoko commented:
Points that still need to be worked on:
- [ ] rendering the pdf when the user tries to sign it (ive already fetched and added the pdf url in the sign component)
- [ ] checking the overall flow of how the signing works
- [ ] generating a receipt that jurre mentioned

DonKoko reviewed (commented):
## Feedback

- [x] Somehow we have an error when i try go create an asset, related to the asset image:
```
{"level":50,"time":1740494033869,"pid":26970,"hostname":"Nikolays-MacBook.local","err":{"cause":{"cause":{"cause":{"cause":{"stack":"Error: Input Buffer is empty\n    at Sharp._createInputDescriptor (/Users/donkoko/devs/shelf/webapp/node_modules/sharp/lib/input.js:51:13)\n    at new Sharp (/Users/donkoko/devs/shelf/webapp/node_modules/sharp/lib/constructor.js:361:29)\n    at Sharp (/Users/donkoko/devs/shelf/webapp/node_modules/sharp/lib/constructor.js:178:12)\n    at Module.cropImage (/Users/donkoko/devs/shelf/webapp/app/utils/crop-image.ts:10:18)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async uploadFile (/Users/donkoko/devs/shelf/webapp/app/utils/storage.server.ts:78:32)\n
```
- [x] Once a template is added, i cant add anymore as there is no "Add template" button
<img width="1410" alt="Screenshot 2025-02-25 at 16 08 07" src="https://github.com/user-attachments/assets/5a3370ff-e357-48e7-bfa9-7f0ff6787930" />

- [ ] When i click the 3 dots on a template to edit it for example, i get the following error:
<img width="622" alt="Screenshot 2025-02-25 at 16 09 58" src="https://github.com/user-attachments/assets/feb91215-0e2b-4485-832b-73e8a2ddb5ee" />

- [x] revision label should not break row
<img width="698" alt="Screenshot 2025-02-25 at 16 14 34" src="https://github.com/user-attachments/assets/a9a1f1aa-ade7-47c3-a5d4-ee368e712810" />

- [x] We should implement the template selector for bulk assign custody as well
- [x] Settings and learn more links should open in new tabs
<img width="418" alt="Screenshot 2025-02-25 at 16 19 10" src="https://github.com/user-attachments/assets/3aaaa8dc-45d1-470a-8199-c76b9994ba9d" />

- [ ] We need to better handle the case when custody is assigned to a User and the URL is visited without having a session(see screenshot). We should allow the user to login and be redirected back to the page or at least have a better message, instead of this generic protected route error.
<img width="901" alt="Screenshot 2025-02-25 at 16 21 24" src="https://github.com/user-attachments/assets/65e12608-850a-4e10-a25a-fa6f777474ef" />


- [ ] The spacing on the templates page seems outdated(larger than other views). Lets update it
<img width="1381" alt="Screenshot 2025-02-25 at 16 30 37" src="https://github.com/user-attachments/assets/2927d581-6aa4-4f0a-908f-1d2efa669913" />

- [ ] We should have a friendlier view when asset custody is already signed, not just the default error. Maybe even a link to the receipt or actually showing the already signed view
<img width="908" alt="Screenshot 2025-02-25 at 16 31 55" src="https://github.com/user-attachments/assets/a16d0398-56b3-4b5c-9143-5f79b0372d95" />

- [ ] When I assign using a template for a second time i get this error:

```
{"level":50,"time":1740494846048,"pid":26970,"hostname":"Nikolays-MacBook.local","err":{"cause":{"name":"PrismaClientKnownRequestError","code":"P2002","clientVersion":"6.4.1","meta":{"modelName":"Asset","target":["templateId"]},"stack":"PrismaClientKnownRequestError: \nInvalid `prisma.asset.update()` invocation:\n\n\nUnique constraint failed on the fields: (`templateId`)\n
```
I believe we discussed this and we decided that we might have to change the DB structure to allow this.

- [ ] Lets disable "Bookings" type when creating a template. I think we should not remove it from the DB, just hardcode it in the form so its always "custody"(for now)

## Questions

- [ ] Once custody is signed for an asset, i am not sure where i can see it and where i can see the signature receipt @carlosvirreira @rockingrohit9639 
- [ ] Should we have an index of signed custody receipts? @carlosvirreira @rockingrohit9639

DonKoko reviewed (commented):
- [x] This error keeps on happening again and prevents me from changing default template
![image](https://github.com/user-attachments/assets/dc58cfbc-5b9d-448e-8339-d05a47598658)
- [x] Placeholders should be changed from Booking to Custody
![image](https://github.com/user-attachments/assets/545003a5-78ff-4e4e-8bb1-0316aff50d66)
- [ ] Update this text to not include "bookings"
![image](https://github.com/user-attachments/assets/0822eef2-192b-4e26-b796-ef61c72948b3)
- [ ] After bulk assigning custody with signature I get a dom nesting error:
![image](https://github.com/user-attachments/assets/8b392485-fb62-4304-a41c-7fd079920f79)
- [ ] Error when opening a signing link for custody assigned to user is still too generic:
![image](https://github.com/user-attachments/assets/b85cc842-0f09-42b9-b2d8-f35c11b746ef)
Lets do this:
```
Title: Not allowed
Message: This custody agreement requires you to be logged in to sign it.
```
Lets add a login button with a `redirectTo` back to this url so the process is smooth. I think we might have to figure out a way to append the login button to the message. so the user can have this smooth UX and not have to manually go back and forth and click the link again.
- [x] When I am on this route `settings/team/users/userId/assets` and i click the icon for signing custody link, i get an 404. We need to make sure this modal can work in all cases.
- [ ] This actions reference is wrong. Its now called "Release custody"
![image](https://github.com/user-attachments/assets/99c0d4eb-0594-4293-9585-8ed7daa09ba6)



CustomTierLimit seems to be causing a lot of issues. Here are some stuff i found:
- [ ] The max templates on custom tier limit and tier limit in general should be per-workspace, not per user. Now i have 3 in custom tier limit and it works in total and not per workspace
- [ ] When the limit is finished, we shouldnt just hide the button. We should handle it like in the other cases when the button is disabled
- [x] However I can go to `/templates/new` and create new one, even if I have reached limit. We need to check that on the server always

DonKoko reviewed (commented):
Overall really great job. We still have some stuff to handle. The points i have not checked from the previous review are still pending.

- [x] Signed custody routes should only be available in team workspaces. Same as bookings:
<img width="264" alt="Screenshot 2025-03-09 at 15 32 57" src="https://github.com/user-attachments/assets/12d14c80-83f8-4fd2-95a2-3e18f8035f3d" />


- [x] Placeholders in the form say "booking" instead of "custody"
<img width="1012" alt="Screenshot 2025-03-09 at 15 38 16" src="https://github.com/user-attachments/assets/bc23297c-dbaf-4566-a600-522fb4a589ca" />

- [x] "Make default" action doesn't have an icon. Lets use the `star` icon from lucide
<img width="698" alt="Screenshot 2025-03-09 at 15 42 49" src="https://github.com/user-attachments/assets/7468651b-7cf8-4f65-97d6-4c8a8f8b8190" />

- [ ] With the new approach we are taking of only being able to change a custody agreement file once, after its used, we dont need the "current revision" tracking and label
<img width="660" alt="Screenshot 2025-03-09 at 15 45 07" src="https://github.com/user-attachments/assets/37ccf48d-7b9c-46c3-9d50-587659eb6676" />

- [ ] Weird error message. Made me laugh
<img width="541" alt="Screenshot 2025-03-09 at 16 01 37" src="https://github.com/user-attachments/assets/1b4eb18e-70c9-46fe-8000-84568ebca202" />

- [ ] Tried releasing custody of assets that already had custody assigned before this PR. Happens both in bulk and singular asset release. I get the following error:
```
{"level":50,"time":1741528872507,"pid":36947,"hostname":"Nikolays-MacBook.local","err":{"cause":{"stack":"TypeError: Cannot read properties of undefined (reading 'id')\n    at eval (/Users/donkoko/devs/shelf/webapp/app/modules/asset/service.server.ts:1938:54)\n    at Array.map (<anonymous>)\n    at eval (/Users/donkoko/devs/shelf/webapp/app/modules/asset/service.server.ts:1937:16)
```

- [ ] Receipts index is missing pagination
<img width="1425" alt="Screenshot 2025-03-09 at 16 21 00" src="https://github.com/user-attachments/assets/d04793f4-36a0-4175-9cfa-f260596b8150" />

- [ ] ON the screen where the user signs a custody, we dont have a success case or anything like this. For users it just redirects to the asset and for NRM it redirects to the login page. I think we need to have a nice success message with some actions. @carlosvirreira can you help with the design on this?

- [ ] The `TierLimit` for agreements should work like with custom fields. It should be based on Active agreements not the total count. LMK if you need help adjusting this but you should be able to use the custom fields as an example.

carlosvirreira commented:
Hi @DonKoko (cc @rockingrohit9639 )

Where a custodian/user signs custody - we have a success case design. Please see below:
![Screenshot 2025-03-10 at 11 50 10](https://github.com/user-attachments/assets/a5bde468-eb43-48c6-94d3-39c1b8459d8c)

I do notice that, for the case of an admin selecting that all the user/nrm has to do is just 'receive' the template, for this we indeed have no success case. The thing here @DonKoko is that, the close action does nothing. We are not storing any consent or anything like this so it made sense to just leave it as it is. On retrospect, if you click close we could show some sort of a 'thank you' page. 

https://www.figma.com/design/ykvc0MhPeNLsrJ4tJsKosQ/Shelf-%7C-UX-Scenarios?node-id=911-31483
![Screenshot 2025-03-10 at 13 17 25](https://github.com/user-attachments/assets/bcedae4e-bc06-4d91-a736-78fc6dd0ea01)

DonKoko reviewed (commented):
- [x] Navigating to the `agreements` and `receipts` urls when in personal allows you to view it. The whole route should be protected not just the menu item. See how we do it in bookings
- [x] This is still the case: Tried releasing custody of assets that already had custody assigned before this PR. Happens both in bulk and singular asset release. I get the following error:
```
{"level":50,"time":1742296975657,"pid":474455,"hostname":"DESKTOP-4O50AD2","err":{"cause":{"stack":"TypeError: Cannot read properties of undefined (reading 'id')\n    at eval (/home/donkoko/devs/shelf/webapp/app/modules/asset/service.server.ts:1935:54)\n    at Array.map (<anonymous>)\n    at eval (/home/donkoko/devs/shelf/webapp/app/modules/asset/service.server.ts:1934:16)\n
```

- [x] Email text needs adjustments. Here is the new text:
```
Howdy,

Don Koko has assigned you as custodian for Beckman Coulter DelsaMax PRO.
Please click the link below to view the custody agreement and sign it:
https://localhost:3000/sign/cm8ef92c4000et53brxtmggi2

To view the asset, please click the link below:
https://localhost:3000/assets/a23g7dcmay2lwxn6zxilaodc6

Thanks,
The Shelf Team
```

- [x] Activating a disabled agreement when limit is reached, should work the same as creating a new one and have the same validations
- [x] When assigning custody to an NRM, the text should be different:
![image](https://github.com/user-attachments/assets/afab353d-b43a-4b8e-beca-2f0dcde09fc4)

```
This PDF agreement page has been published. Chief Surveyor will be able to visit this page to read (and sign) the document. Make sure you send them the link. You can visit the asset page to open this modal in case you need to acquire the share link again
```

- [x] Something weird happened here with the vertical spacing between name and label
![image](https://github.com/user-attachments/assets/205debdf-5c73-4414-9b09-1714625a347e)

- [x] When assigning custody to a User that doesn't require signature, we should still send a link to preview the PDF in the email. Now they cant see the PDF in any way. We should have a design for this. If you cant find it, pls ask on Discord
- [x] Related to the previous point, when assigning to NRM with agreement that doens't require signature, we are showing a link where we are showing the signature feature. I think both this and the case above should be handled the same. The view should be the same just without the signature part. We should have a design for this.

- [ ] When assigning custody to User that doesnt require signature, its being displayed weird in the receipts:
![image](https://github.com/user-attachments/assets/cccd3643-56cc-4ea0-bd76-3bc6d035ad99)

Lets do the following:
1. Signature status - Not required
2. Custody status - should be the same as in the other cases. IN my screenshot it should be in progress

DonKoko commented:
@rockingrohit9639 yesterday we discussed the last pending point. Please find the information below on what needs to be implemented:

### Limits
- 5MB max file size for custody agreement PDF
- Max 5 active agreements per workspace
- Max 20 total agreements per workspace - this is a new thing but we need to do it in order to limit the max storage space used

### User safety
To ensure user safety and prevent abuse from admins we have to make it so that once a template has been used once(signed or unsigned), we should not allow to change the file. To do that please disable the file field in that case with the following message: `This custody agreement has already been used and signed and it's file cannot be changed anymore. If your agreement document has changed, please create a new custody agreement with your new document.`

DonKoko reviewed (commented):
- [x] Custody agreements limits should be 0 for free tier and tier_1. Only tier_2(team) can have them
- [ ] Adding a new agreement seems to check `maxCustodyAgreements` instead of `maxActive...`. I could add a 6th one:
![Screenshot 2025-03-20 at 16 54 41](https://github.com/user-attachments/assets/2c01232d-55bc-415a-90e2-e71965f21e45)
- [ ] When custody is assigned with agreement that doesnt require signature, the link to the agreement should still be in the email.
- [ ] Editing a file from an agreement that is used works good, but i dont see the message: `This custody agreement has already been used and signed and it's file cannot be changed anymore. If your agreement document has changed, please create a new custody agreement with your new document.`

carlosvirreira commented:
**Comments/Feedback:**

**Template upload area**
Please add supportive text.
![Screenshot 2025-03-25 at 18 14 59](https://github.com/user-attachments/assets/cdb54565-22e6-4ae8-a781-d2632ac2a29f)

Please add a column to show if the agreement is 'Ask for signature' or 'View only' (You can use the same 'pill' - as we have used on the 'default' column. 
![Screenshot 2025-03-25 at 18 18 32](https://github.com/user-attachments/assets/b5cae97a-7da1-44c1-9c0a-32066cf500c2)

**Asset Index**
If you are assigning custody via a batch action (for a SINGULAR) asset. Please emulate the same UX as we have on the asset page when performing this action.  (Right now you have to open the asset page and get this done).

**Advanced Asset Index**
Add same icon to advanced asset index with the same hover text
![Screenshot 2025-03-25 at 18 31 26](https://github.com/user-attachments/assets/bd2d3c4e-b141-4ed1-9f75-cee2402c0050)

**Asset Page**
BUG (Although the asset is signed for ((in custody)) you can still see the 'awaiting signature' component.
![Screenshot 2025-03-25 at 18 33 30](https://github.com/user-attachments/assets/6826037a-af71-40ac-a8e5-73a57340fe30)

**Receipts page**
Add filters. Custody and Signature Status. Same style as 'asset status filters'

**Receipt explorer/modal**
- Generate a PDF instead of a JPG
- Since the asset name can change. Print the asset ID under the asset name

** Signature collection screen **
Please add a 'clear' that deletes any sort of signature that is on the view.
![Screenshot 2025-03-25 at 18 39 41](https://github.com/user-attachments/assets/85f1af38-ce35-4693-9fb8-3a6bba208d24)

Please make the signature font look signature-ish (https://github.com/documenso/documenso/tree/main/apps/remix/public/fonts) use Caveat.

** View Only Agreement **

Please add a 'close' button in the same style as the 'sign' button. As users might be confused on what to do on that view.

carlosvirreira commented:
Additional tasks:

**When using Bulk Assign Custody to a Self Service OR Base user with a Sign action**
After the batch action is performed please show a message.
![Screenshot 2025-03-26 at 18 11 20](https://github.com/user-attachments/assets/0f8b4764-d920-4582-9c45-effcc6f2cb90)

**When using Bulk Assign Custody to a NRM with a Sign action**
![Screenshot 2025-03-26 at 18 12 28](https://github.com/user-attachments/assets/4a636cb1-1940-4a9b-a264-dc19535d4a05)

**When using Bulk Assign Custody to a NRM user with a View Only Agreement**
![Screenshot 2025-03-26 at 18 16 01](https://github.com/user-attachments/assets/989aa7f2-f2fc-49fe-973f-543e3a1d0004)

**When using Bulk Assign Custody to a Self Service OR Base user with a View Only Agreement**
![Screenshot 2025-03-26 at 18 23 19](https://github.com/user-attachments/assets/1e6926c4-9362-48bd-b46b-b30b328454d7)

DonKoko reviewed (commented):
- [ ] @carlosvirreira we still need links for the name and description "See sample" in the new agreement form

- [x] The more i test this the more i think that the concept of having the status available while waiting for signature is not good. I find it very confusing. I think we should consider doing something different here. @rockingrohit9639 could you implement a front-end only solution for now? You can see i did something similar in the booking-status-badge. So a few things: 1) We should actually have AwaitingSignatureTooltip inside the asset-status-badge component. 2) If `CustodySignatureStatus.PENDING` we can show in the status badge "Signature pending" with a yellow color(like our normal warning color)

- [x] This is just partially fixed:
> Asset Page
> BUG (Although the asset is signed for ((in custody)) you can still see the 'awaiting signature' component.

![Screenshot 2025-04-02 at 15 42 00](https://github.com/user-attachments/assets/93e10839-10c4-4898-85fe-e01fccb2547c)
Lets make it so once its signed, the link says "View receipt". At this point it doesn't make sense to share the document again.

- [x] On the receipts index, the filters and the search have some weird styling. Fonts are wrong size and weight as well as spacings in the dropdowns. Please make them the same as on the simple asset index filters.
![Screenshot 2025-04-02 at 15 43 44](https://github.com/user-attachments/assets/d6cae69a-4ab2-4931-870f-f5bb877185fd)

- [ ] @carlosvirreira i think we have a big gap in this whole PR, concerning Kits. Kits have the same custody functionality and they are widely used by our users, so not having this functionality for kits seems like a huge gap in what our users will expect to see.

DonKoko commented:
Okey so we have the situation about how kits and assets custodies are interacting with each other. Having the signed custody makes this even more complex so we have to take some decisions. I have attempted to make it in the most simplified and easy way to manage, that also leaves room for future improvement and extending of the feature. This is now becoming a rework of custodies that I think will be very beneficial in the long run.

## Goals

1. Give enough user agency to the admins so they feel they are not limited by the software, but also they are not let down and can trust the software to "protect" them when needed
2. Clearly communicate the consequences of actions that mix custody of singular assets and kits


## Cases

### Assigning custody of a kit

1. When an asset or multiple assets that are part of a Kit are already  `In Custody` or `Signature pending` individually. The kit custody assignment should be allowed. When the modal opens we need to change some working and clearly communicate what is going to happen:
  - [ ] There should be a warning message explaining the following: "The kit already has some assets in custody. My assigning custody to the kit, the asset custody will be synchronized with the kit. All current custody or pending custody of the assets will be released and they will get assigned custody via the kit."
  - [ ] We should add an extra confirmation dialog where the user confirms this action
  
  
 ### Assigning custody to an asset that is part of a kit
 
 1. When the kit is in custody, assigning or releasing individual custody to the asset should not be allowed. To achieve that the user should just remove the asset from the kit and handle it separately.
 2. When the kit is `Available`, assigning individual custody of an asset should be allowed. This already works, however we should however add a warning message in the custody assignment modal: "This asset is part of a kit. By assigning it individual custody, you might get some inconsistent information and face limitations when trying to update the kit custody later on." This should be the case for both signed and normal custody.
 
 
 ### Assigning custody to a kit with a signature requirement

1. When the status of the kit is `SIGNATURE_PENDING` the status of the assets should also be `SIGNATURE_PENDING`
2. Individual assets that have `SIGNATURE_PENDING` via a kit, should have a bit of different UI and some different messages:
  - [ ] The status badge text should change to the following in this case: "Asset status will change to "In custody" after signing. The custody of this asset has been assigned via a Kit, so you cannot currently release its custody individually. If you want to make it available again, without affecting the Kit custody, you need to remove this asset from the kit."
  - [ ] The release custody button of the asset should be completely disabled in this case, with the same tooltip: "The custody of this asset has been assigned via a Kit, so you cannot currently release its custody individually. If you want to make it available again, without affecting the Kit custody, you need to remove this asset from the kit from the kit page."
  - [ ] On the right sidebar, of the assets, we should show the same card as usual, however with a bit different text: `Awaiting signature from ${user.name}. Custody assigned via ${kit.name}`.  The kit name should be a new tab link to the kit
  
  
### Asset Status badge improvements
  
- [ ] When an asset is in custody via a kit, the badge should say "In custody via kit" and clicking it should show a tooltip with the following content:
```
In custody via kit
This asset has been assigned custody via a kit: ${kit.name}.
````

- [ ] When an asset has a status `SIGNATURE_PENDING` because of the parent kit, visually we should display "Signature Pending via Kit". There should also be a tooltip:
```
In custody via kit
This asset has been assigned custody via a kit: ${kit.name}. The custody is still awaiting signature. To find more information go to the kit page.
```


> [!CAUTION]
> This is everything I could think of so far.  I can't be 100% confident i havent missed very important cases, but I think this is a good starting point.

DonKoko reviewed (commented):
- [x] Clicking badge with status `SIGNATURE_PENDING` of an asset that was assigned custody as an individual asset, but is part of a kit, gives an error:
![Screenshot 2025-04-22 at 14 52 13](https://github.com/user-attachments/assets/d7b41394-fb0d-4afd-b6fd-1b6a27e9b6d7)

- [ ] When a kit has custody, if I navigate to the page of an asset that is part of this kit, and click on actions, i can see that "Release custody" is disabled visually. But I can still click it and open the modal and release its custody individually. This should not be the case

- [ ] When assigning custody to a kit with required signature, i didnt receive the email. Also in the server logs i dont see the email being sent

- [ ] Receipts pagination doesnt work. I cant go to page 2
![Screenshot 2025-04-22 at 16 01 32](https://github.com/user-attachments/assets/64776971-a50f-49c6-bbf9-8c0e730b7dd6)

- [ ] Clicking on a receipt of a kit, updates the url but doesnt show the modal. Here is how the url looked like: https://localhost:3000/receipts?per_page=100&receiptId=cm9siffoq001homaqnbnhte4d but the modal was not showing

DonKoko commented:
@rockingrohit9639 we discussed everything and its approved. You can continue with implementing the suggestions I gave & the review comments.

DonKoko reviewed (commented):
- [ ] When i try to assign custody of a kit to an NRM without an agreement, the transaction fails. This happens to a kit that has 13 assets inside it. I also tried it with another kit that has 4 assets inside it and it worked. So something is not right with our query. Its a simple request that shouldnt be failing. Maybe something is inside the transaction that shouldn't be there.
```
{"level":50,"time":1746447183155,"pid":67546,"hostname":"Nikolays-MacBook.local","err":{"cause":{"code":"P2028","meta":{"modelName":"Asset","error":"Transaction already closed: A batch query cannot be executed on an expired transaction. The timeout for this transaction was 5000 ms, however 5035 ms passed since the start of the transaction. Consider increasing the interactive transaction timeout or doing less work in the transaction."},"clientVersion":"6.7.0","name":"PrismaClientKnownRequestError","stack":"PrismaClientKnownRequestError: \nInvalid `prisma.asset.update()` invocation:\n\n\nTransaction API error: Transaction already closed: A batch query cannot be executed on an expired transaction. The timeout for this transaction was 5000 ms, however 5035 ms passed since the start of the transaction. Consider increasing the interactive transaction timeout or doing less work in the transaction.\n
```

- [ ] Currently the receipts index also shows receipts for custodies that don't have an agreement attached to them. This should not be the case, those ones should not be shown there. Only show custodies that have an agreement attached
- [ ] When clicking the receipt that doesnt require signature it says `not_required`
<img width="778" alt="Screenshot 2025-05-05 at 15 20 45" src="https://github.com/user-attachments/assets/c8f86d4a-2894-47d2-b542-40b34fa34adb" />


- [ ] I had a kit with an asset inside it that had "Signature pending". I assigned custody to the kit and the status got synchronized. All worked well, except when I look at the receipts I still see the asset custody as pending and it should be cancelled. The receipts should also be updated in this case cancelled as the custody was cancelled by assigning the kit custody
<img width="1399" alt="Screenshot 2025-05-05 at 15 36 18" src="https://github.com/user-attachments/assets/13816692-b2e3-4565-a570-81710094c6dc" />

rockingrohit9639 commented:
Hi @DonKoko 
> When i try to assign custody of a kit to an NRM without an agreement, the transaction fails. This happens to a kit that has 13 assets inside it. I also tried it with another kit that has 4 assets inside it and it worked. So something is not right with our query. Its a simple request that shouldnt be failing. Maybe something is inside the transaction that shouldn't be there.

I was not able to reproduce this error on my side. I tried assigning custody to a kit which had 15 assets and it still worked somehow.

DonKoko commented:
@rockingrohit9639 Can you please look at this PR #1777 
Its a change that requires to be integrated with this PR. There are also conflicts related to it. Could you please resolve them.

carlosvirreira commented:
**Signed Custody (https://testapp.shelf.nu/agreements)**
- [ ] If you edit an agreement you will see some links that read "See a sample" under Name* and Description. These are leading nowhere. 

**Assigning Custody of a Kit (https://testapp.shelf.nu/kits)**
Scenario tested: Workspace Admin -> Base User
Signature required? : Yes
All assets available and no issues with kit?: Yes
- [ ] I do not get a notification (email) once my requested signer has signed the agreement, this would be nice so I can be informed. I have to now basically go to the receipts tag or remember which agreement i had sent to check if it is in custody or not.

Scenario tested: Workspace Admin -> Base User
Signature required? : No
All assets available and no issues with kit?: Yes

- [ ] I can assign custody and inmediately I see 'in custody', then a widget that says 'Agreement name' + View receipt. If I click on it I see an old 'signed' custody agreement instead of the 'view only' non required to sign one. This seems to be a bug.  &#8595;

![Screenshot 2025-05-22 at 12 26 48](https://github.com/user-attachments/assets/797fc3ab-e14b-44c3-a369-8e369c2d9096)

![Screenshot 2025-05-22 at 12 27 24](https://github.com/user-attachments/assets/fda40113-3ad7-4ff8-9d8c-abdea667cd08)


**Receiving Custody of a Kit (https://testapp.shelf.nu/kits)**
Scenario tested: **Base user** -> Assigned custody of a kit by Admin
Signature required? : Yes

I log into Shelf and I can see an asset, it reads 'Signature Pending via Kit' > **This signature request intended for me/my user**. When I click on the asset, I go to the asset page. There I see that the asset is awaiting signature from (base test) Custody assigned via kit name. and a 'Share document' link. 

-> If possible, we could change the wording. "Awaiting your signature" "Custody assigned via _kit name_" "Sign document"

* Right now if I click on Share document I can send a mail again or copy the share link. I can (while being logged in) copy the share URL on my browser and open the link to sign it. - Although I can see the agreement and sign it - if i do so I will get an error. The only way to successfully sign a document is to click on the link delivered per mail. 

* On signature success, I get presented with two options. 1) See kit | 2) Go to Dashboard. Clicking on 'Go to Dashboard' will result in an 'Unauthorised'  screen. (You have no permission to perform this action). This should be changed somehow. 

* Kit page layout broken on Desktop. &#8595;
![Screenshot 2025-05-22 at 11 54 14](https://github.com/user-attachments/assets/87b57e0e-6483-4dd9-9d0e-d80b808f4299)

* After I signed for the kit. I can see a widget "Custody agreement 2025" and "View Receipt" if I click on View Receipt I receive an error. If I should not be able to see the receipt let's please remove that. We could load the widget the admin sees (In custody of user name | since (date))

* I can see a menu item called 'Signed custody' and a sub-tab called 'Receipts' if I click it I get an 'Unauthorized, you have no permission to perform this action' this although I have already signed an agreement. 

----------

Scenario tested: **Base user** -> Assigned custody of a kit by Admin
Signature required? : No

- [ ] I get a mail from Shelf that tells me that I have been assigned custody over a kit. The information shown there is not correct. It says that (I - base user) have assigned myself as a custodian for kit name. In this scenario, the administrator (Adminshelfteamworkspace) is the one that assigned custody to me. 

 ![Screenshot 2025-05-22 at 12 33 36](https://github.com/user-attachments/assets/b9dd9a9c-70ea-479c-b37e-ef7c87a91e5c)

- [ ] I can only see the agreement if I click on the link of the mail
- [ ] When I close the agreement page. I get a success. It has two buttons 'To Dashboard | To Kit's Page' > clicking on ' To Dashboard ' will result in an Unauthorized error with a trace ID. 


**General UI Comments**
- [ ] When selecting an agreement, it is not clear what is a signable or view only agreement. My workaround is to add (view only) to the agreement name to get this information.

DonKoko reviewed (commented):
>  As a Self Service user I cannot take custody of an asset. The system is expecting me to select a custodian before I can add a PDF agreement and this is not possible. This happens on take custody of asset/kit.
- [ ] This is still the case. If I click "Take custody" as self service I see this. I suspect it might be related to the new permissions we implemented for self service and base.
![Screenshot 2025-06-11 at 13 01 21](https://github.com/user-attachments/assets/5170fd67-100c-4151-a090-6873405114df)

- [ ] I am self service and there is an asset Pending Signature from me. I can see the asset when i go to its page, but I cant see this asset on the index
![Screenshot 2025-06-11 at 13 05 53](https://github.com/user-attachments/assets/42262249-2130-4208-aaba-a07dd1cdd19c)
![Screenshot 2025-06-11 at 13 06 47](https://github.com/user-attachments/assets/aaa230b4-1d84-40fc-9451-f628a5647822)

- [ ] On the asset index I dont see "Signature pending via kit". This only happens on simple mode. On advanced mode it works as expected.
![Screenshot 2025-06-11 at 13 20 46](https://github.com/user-attachments/assets/07623368-9556-4157-836b-12c0c6ae5a67)
